utils/testgen.py

```diff
@@ -237,6 +237,11 @@ def pytest_generate_tests(metafunc):
             # Skip unwanted types
             continue
 
+        if hasattr(metafunc.function, 'meta'):
```

Comment:
This block definitely needs a comment explaining its purpose, and if possible we should log why we didn't generate tests in this case. Also, I think we need to check specifically for `test_flag` in `meta` before doing the set intersection, since a test with metadata and without test flags could be skipped for no apparent reason.

```python
# names are cheap :)
meta = getattr(metafunc.function, 'meta', None)
test_flags = getattr(meta, 'kwargs', {}).get('test_flags', [])
# If there are test flags, but the intersection of this provider's test flags
# with the test's flags is empty, don't generate tests for this provider
if test_flags and not set(data.get('flags', []) & set(test_flags):
    continue
```